README.md
# iroffer

iroffer is a software program that acts as a fileserver for IRC. It is similar to a FTP server or WEB server, but users can download files using the DCC protocol of IRC instead of a web browser.

## Why use iroffer?

- extremely fast dcc transfers
- extremely low cpu usage, and reasonable ram usage
- its a program not a script that is slowed by a bulky irc program
- only completed transfers are counted
- supports dcc resume
- set max amount of transfers per hostname
- user friendly error messages for users (no "clamp timeout")
- allows a pack to be designated as a "high demand" pack which can have special limitations and its own separate queue
- supports virtual hosts
- auto-send feature, send a pack to someone when they say something
- auto-saves xdcc information
- remote administration via /msg or DCC CHAT
- bandwidth monitoring, shows last 2 minutes bandwidth average
- Allow sending of queued packs when using low amounts of bandwidth, comes in handy when all slots are filled with people transferring 1k/sec, will keep sending out queued items while bandwidth usage is under a specified amount
- Background or Foreground mode. background mode does not require screen and can be run by cron
- Chroot support (run iroffer from inside a chroot'ed environment)
- overall and pack minspeed
- maximum bandwidth limiting, when set, iroffer will not use more than the set amount of bandwidth (keeps your sysadmin happy)
- can set different maximum bandwidth limits depending on time of day and day of week (keeps your sysadmin very happy)
- logging
- auto-ignores flooders
- support for direct, bnc, wingate, and custom proxy irc server connections
- ignore list

## Getting Started

### Supported Platforms

- Linux
- FreeBSD/OpenBSD/NetBSD
- macOS
- Windows (through the WSL)

### Prerequisites

- crypt
- nss_files
- nss_dns

### Building

Run the *configure* script followed by the generated *Makefile*.

No errors or warnings should appear when compiling.

A sample config file is provided.

### How to let colors work while using screen

create a file in your home directory named ".screenrc", and put the
following lines in it:

```sh
termcap  vt100 'AF=\E[3%dm:AB=\E[4%dm'
terminfo vt100 'AF=\E[3%p1%dm:AB=\E[4%p1%dm'
```

### How To Use Cron

Edit the *iroffer.cron* file's *iroffer_dir*, *iroffer_exec*, and
*iroffer_pid* variables

then crontab -e and place the following line in the editor

\*/5 \* \* \* \* /full/path/to/iroffer/iroffer.cron

### Signal Handling

*iroffer* will handle the following signals:

- SIGUSR1 (kill -USR1 xxxx)  jumps to another server (same as admin command "jump")
- SIGUSR2 (kill -USR2 xxxx)  re-reads config file    (same as admin command "rehash")
- SIGTERM (kill xxxx)        shuts down iroffer      (same as admin command "shutdown")

## Acknowledgments

*iroffer* was originally written by PMG (1998-2005) and the projects former home was iroffer.org


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/), but this project **does not** adhere to Semantic Versioning.

## [Unreleased]

[Unreleased]: https://github.com/antiz/iroffer/compare/6edad52d28e6295ec136674348ee5f007f08e206...HEAD

**v1.5.0 includes all code changes from v1.4.1. Those changes won't be listed here again.** *Check below for the v1.4.1 changelog.*

### Changed

- **Breaking: The configure script is now lower case: Configure -> configure**
- Increased and modernized POSIX compatibility.
- New MD5 implementation

### Fixed

- Failure to read 'last log rotate' timestamp from statefile on systems where time_t is 64 bit
- Incorrect printf format specifiers
- Typos in debug output

## ~~[1.4.1] - 2018-01-12~~ [REMOVED]

[1.4.1]: https://github.com/antiz/iroffer/compare/v1.4.b03-orig...6edad52d28e6295ec136674348ee5f007f08e206

*This release was removed because of issues with the version string format of PMG-provided beta releases vs this repositories' release builds. See [antiz/iroffer issue #1]*

[antiz/iroffer issue #1]: https://github.com/antiz/iroffer/issues/1

### Changed

- Reworked the changelog and readme.
- Restructured the projects file layout
- Code clean-ups
- Relicense under the GPL-3.0+
- The *dynip* script can now use either *links*, *lynx* or *wget*.

### Deprecated

- Platform support for *SunOS/Solaris*, *HP-UX*, *IRIX/IRIX64*, *Digital UNIX*, *Mac OS X Server*, *AIX* and *Cygwin*.
- *convertxdccfile*, since the *statefile* has existed for nearly 14 years now.

### Fixed

- Use correct "printf format string" length specifier for size_t and ssize_t.
- Merged Debian patches to fix typos and the *Configure* script.

## 1.4.b03 - 2005-12-12

### Added

- Add *logmessages* config item, similar to *lognotices*.
- Add *xdcclistfileraw* config, includes control characters to XDCC file.

### Changed

- Minor performance changes to memory allocator.
- Better detection of valid usermodes, nick prefixes, and channel-modes.
- Track usermodes of users in channels (shown in **CHANL** command).
- *clearrecords* should clear the transfer limit too.
- Mention directories when you do an **ADDDIR** or **ADDNEW** command
- Some cleanups to *Configure* and *Makefile*.

### Fixed

- Fix invalid CTCP ping response.
- Fix username when it cannot be determined automatically.
- Fix infinite loop when debug output causes memory allocation.

---

## *Legacy Changelog*

## version 1.4

## 1.4.b02 - 2005-01-16

- all fixes in 1.3.b10
- add autoignore_threshold - auto-ignore rate configurable
- add 'addnew' command

## 1.4.b01 - 2004-09-01

- add autoignore_exclude configurable
- wrap transfer id at 999 or more
- silence file has changed warning for xdcc file
- don't allow a flood of xdcc lists to queue up when we are already busy
- use slow notice for flood
- add server password support
- add daily/weekly/monthly transfer limits

## version 1.3

## 1.3.b11 - 2005-12-12

- fix invalid CTCP ping response
- fix username when it cannot be determined automatically
- fix infinite loop when debug output causes memory allocation

## 1.3.b10 - 2005-01-16

- ignore dev/inode changes as file changes
- add sanity check to 'add' command to only allow regular files
- reaping of old listen ports wasn't very good, caused small leaks
- display memstat list using copy of list
- improve dns lookup error reporting
- fix memory leak with regcomp
- allow 5 minute grace period before disconnecting due to restrictsend
- windows users like their \r characters in xdcclistfile

## 1.3.b09 - 2004-08-23

- fix build issues with OSF1
- fix unaligned bus fault for 64bit writes in md5sum state structure
- fix infinite loop when more than 1 transfer in progress
- don't allow plistoffset must be greater than plist time

## 1.3.b08 - 2004-08-09

- don't use in_port_t, it's not in older glibc's
- fix warnings
- make transfer method determined at runtime
- mmap doesn't work on cygwin sometimes
- warn users about stupid usenatip values
- don't use tr_id 0
- fix crash if rmq called with invalid argument
- fix build for macosx

## 1.3.b07 - 2004-04-06

- add convertxdccfile (convert .xdcc -> .state)
- don't ignore nickserv, re-register on successful nick rename

## 1.3.b06 - 2004-04-03

- make debug a cmdline option not config file
- replace auto* with autosendpack
- change overallminspeed -> transferminspeed
- remove double queue, will add something better in the future
- add xdcclistfile config, save xdl to text file
- update info and xdl commands, add 'xdcc info' message for users
- add nomd5sum config option
- add md5sum calculation to packs
- fixup logstats
- add removedir, remove all files from within a directory
- share filedescriptors and mmaps between transfers
- fix total sent problem in statefile
- fix integer overflow if transfer rate exceeds 140mbps
- add sendfile support (for linux and freebsd)
- turn off respondtochannellist by default
- improve log rotation
- fix 'shutdown now' crash if issued through dcc chat
- other cleanups

## 1.3.b05 - 2004-03-22

- dcc chat fixes
- stop updating context log after crash starts
- fix non-blocking connect error checking
- change some privmsg to notice
- fix various rehash crashes if items removed/added
- add generic state file (remove xdccfile, messagefile, ignorefile)
- fix user xdcc list to use notice, broke a while back
- fix error handling if can't bind to listen socket
- track nicks and close when no longer on channels (if restrictsend)
- add punishslowusers

## 1.3.b04 - 2004-02-22

- fix crash on shutdown

## 1.3.b03 - 2004-02-21

- less calls to gethostbyname()
- do server dns resolve in a child process
- fix solaris build
- create fast server queue
- dcc chat cleanup, make connection non-blocking
- add 'chatl' and 'closec'
- multiple dcc chat support
- fix VERASE and VWERASE

## 1.3.b02 - 2004-01-22

- fix build for MacOSX and *BSD

src/iroffer_md5.c
/*
 * This is an OpenSSL-compatible implementation of the RSA Data Security, Inc.
 * MD5 Message-Digest Algorithm (RFC 1321).
 *
 * Homepage:
 * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
 *
 * Author:
 * Alexander Peslyak, better known as Solar Designer <solar at openwall.com>
 *
 * This software was written by Alexander Peslyak in 2001.  No copyright is
 * claimed, and the software is hereby placed in the public domain.
 * In case this attempt to disclaim copyright and place the software in the
 * public domain is deemed null and void, then the software is
 * Copyright (c) 2001 Alexander Peslyak and it is hereby released to the
 * general public under the following terms:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 *
 * (This is a heavily cut-down "BSD license".)
 *
 * This differs from Colin Plumb's older public domain implementation in that
 * no exactly 32-bit integer data type is required (any 32-bit or wider
 * unsigned integer data type will do), there's no compile-time endianness
 * configuration, and the function prototypes match OpenSSL's.  No code from
 * Colin Plumb's implementation has been reused; this comment merely compares
 * the properties of the two independent implementations.
 *
 * The primary goals of this implementation are portability and ease of use.
 * It is meant to be fast, but not as fast as possible.  Some known
 * optimizations are not included to reduce source code size and avoid
 * compile-time configuration.
 */

#ifndef HAVE_OPENSSL

#include <string.h>

#include "iroffer_md5.h"

/*
 * The basic MD5 functions.
 *
 * F and G are optimized compared to their RFC 1321 definitions for
 * architectures that lack an AND-NOT instruction, just like in Colin Plumb's
 * implementation.
 */
#define F(x, y, z) ((z) ^ ((x) & ((y) ^ (z))))
#define G(x, y, z) ((y) ^ ((z) & ((x) ^ (y))))
#define H(x, y, z) (((x) ^ (y)) ^ (z))
#define H2(x, y, z) ((x) ^ ((y) ^ (z)))
#define I(x, y, z) ((y) ^ ((x) | ~(z)))

/*
 * The MD5 transformation for all four rounds.
 */
#define STEP(f, a, b, c, d, x, t, s)                                           \
    (a) += f((b), (c), (d)) + (x) + (t);                                       \
    (a) = (((a) << (s)) | (((a)&0xffffffff) >> (32 - (s))));                   \
    (a) += (b);

/*
 * SET reads 4 input bytes in little-endian byte order and stores them in a
 * properly aligned word in host byte order.
 *
 * The check for little-endian architectures that tolerate unaligned memory
 * accesses is just an optimization.  Nothing will break if it fails to detect
 * a suitable architecture.
 *
 * Unfortunately, this optimization may be a C strict aliasing rules violation
 * if the caller's data buffer has effective type that cannot be aliased by
 * MD5_u32plus.  In practice, this problem may occur if these MD5 routines are
 * inlined into a calling function, or with future and dangerously advanced
 * link-time optimizations.  For the time being, keeping these MD5 routines in
 * their own translation unit avoids the problem.
 */
#if defined(__i386__) || defined(__x86_64__) || defined(__vax__)
#define SET(n) (*(MD5_u32plus*)&ptr[(n)*4])
#define GET(n) SET(n)
#else
#define SET(n)                                                                 \
    (ctx->block[(n)] = (MD5_u32plus)ptr[(n)*4] |                               \
                       ((MD5_u32plus)ptr[(n)*4 + 1] << 8) |                    \
                       ((MD5_u32plus)ptr[(n)*4 + 2] << 16) |                   \
                       ((MD5_u32plus)ptr[(n)*4 + 3] << 24))
#define GET(n) (ctx->block[(n)])
#endif

/*
 * This processes one or more 64-byte data blocks, but does NOT update the bit
 * counters.  There are no alignment requirements.
 */
static const void* body(MD5_CTX* ctx, const void* data, unsigned long size) {
    const unsigned char* ptr;
    MD5_u32plus a, b, c, d;
    MD5_u32plus saved_a, saved_b, saved_c, saved_d;

    ptr = (const unsigned char*)data;

    a = ctx->a;
    b = ctx->b;
    c = ctx->c;
    d = ctx->d;

    do {
        saved_a = a;
        saved_b = b;
        saved_c = c;
        saved_d = d;

        /* Round 1 */
        STEP(F, a, b, c, d, SET(0), 0xd76aa478, 7)
        STEP(F, d, a, b, c, SET(1), 0xe8c7b756, 12)
        STEP(F, c, d, a, b, SET(2), 0x242070db, 17)
        STEP(F, b, c, d, a, SET(3), 0xc1bdceee, 22)
        STEP(F, a, b, c, d, SET(4), 0xf57c0faf, 7)
        STEP(F, d, a, b, c, SET(5), 0x4787c62a, 12)
        STEP(F, c, d, a, b, SET(6), 0xa8304613, 17)
        STEP(F, b, c, d, a, SET(7), 0xfd469501, 22)
        STEP(F, a, b, c, d, SET(8), 0x698098d8, 7)
        STEP(F, d, a, b, c, SET(9), 0x8b44f7af, 12)
        STEP(F, c, d, a, b, SET(10), 0xffff5bb1, 17)
        STEP(F, b, c, d, a, SET(11), 0x895cd7be, 22)
        STEP(F, a, b, c, d, SET(12), 0x6b901122, 7)
        STEP(F, d, a, b, c, SET(13), 0xfd987193, 12)
        STEP(F, c, d, a, b, SET(14), 0xa679438e, 17)
        STEP(F, b, c, d, a, SET(15), 0x49b40821, 22)

        /* Round 2 */
        STEP(G, a, b, c, d, GET(1), 0xf61e2562, 5)
        STEP(G, d, a, b, c, GET(6), 0xc040b340, 9)
        STEP(G, c, d, a, b, GET(11), 0x265e5a51, 14)
        STEP(G, b, c, d, a, GET(0), 0xe9b6c7aa, 20)
        STEP(G, a, b, c, d, GET(5), 0xd62f105d, 5)
        STEP(G, d, a, b, c, GET(10), 0x02441453, 9)
        STEP(G, c, d, a, b, GET(15), 0xd8a1e681, 14)
        STEP(G, b, c, d, a, GET(4), 0xe7d3fbc8, 20)
        STEP(G, a, b, c, d, GET(9), 0x21e1cde6, 5)
        STEP(G, d, a, b, c, GET(14), 0xc33707d6, 9)
        STEP(G, c, d, a, b, GET(3), 0xf4d50d87, 14)
        STEP(G, b, c, d, a, GET(8), 0x455a14ed, 20)
        STEP(G, a, b, c, d, GET(13), 0xa9e3e905, 5)
        STEP(G, d, a, b, c, GET(2), 0xfcefa3f8, 9)
        STEP(G, c, d, a, b, GET(7), 0x676f02d9, 14)
        STEP(G, b, c, d, a, GET(12), 0x8d2a4c8a, 20)

        /* Round 3 */
        STEP(H, a, b, c, d, GET(5), 0xfffa3942, 4)
        STEP(H2, d, a, b, c, GET(8), 0x8771f681, 11)
        STEP(H, c, d, a, b, GET(11), 0x6d9d6122, 16)
        STEP(H2, b, c, d, a, GET(14), 0xfde5380c, 23)
        STEP(H, a, b, c, d, GET(1), 0xa4beea44, 4)
        STEP(H2, d, a, b, c, GET(4), 0x4bdecfa9, 11)
        STEP(H, c, d, a, b, GET(7), 0xf6bb4b60, 16)
        STEP(H2, b, c, d, a, GET(10), 0xbebfbc70, 23)
        STEP(H, a, b, c, d, GET(13), 0x289b7ec6, 4)
        STEP(H2, d, a, b, c, GET(0), 0xeaa127fa, 11)
        STEP(H, c, d, a, b, GET(3), 0xd4ef3085, 16)
        STEP(H2, b, c, d, a, GET(6), 0x04881d05, 23)
        STEP(H, a, b, c, d, GET(9), 0xd9d4d039, 4)
        STEP(H2, d, a, b, c, GET(12), 0xe6db99e5, 11)
        STEP(H, c, d, a, b, GET(15), 0x1fa27cf8, 16)
        STEP(H2, b, c, d, a, GET(2), 0xc4ac5665, 23)

        /* Round 4 */
        STEP(I, a, b, c, d, GET(0), 0xf4292244, 6)
        STEP(I, d, a, b, c, GET(7), 0x432aff97, 10)
        STEP(I, c, d, a, b, GET(14), 0xab9423a7, 15)
        STEP(I, b, c, d, a, GET(5), 0xfc93a039, 21)
        STEP(I, a, b, c, d, GET(12), 0x655b59c3, 6)
        STEP(I, d, a, b, c, GET(3), 0x8f0ccc92, 10)
        STEP(I, c, d, a, b, GET(10), 0xffeff47d, 15)
        STEP(I, b, c, d, a, GET(1), 0x85845dd1, 21)
        STEP(I, a, b, c, d, GET(8), 0x6fa87e4f, 6)
        STEP(I, d, a, b, c, GET(15), 0xfe2ce6e0, 10)
        STEP(I, c, d, a, b, GET(6), 0xa3014314, 15)
        STEP(I, b, c, d, a, GET(13), 0x4e0811a1, 21)
        STEP(I, a, b, c, d, GET(4), 0xf7537e82, 6)
        STEP(I, d, a, b, c, GET(11), 0xbd3af235, 10)
        STEP(I, c, d, a, b, GET(2), 0x2ad7d2bb, 15)
        STEP(I, b, c, d, a, GET(9), 0xeb86d391, 21)

        a += saved_a;
        b += saved_b;
        c += saved_c;
        d += saved_d;

        ptr += 64;
    } while (size -= 64);

    ctx->a = a;
    ctx->b = b;
    ctx->c = c;
    ctx->d = d;

    return ptr;
}

void MD5_Init(MD5_CTX* ctx) {
    ctx->a = 0x67452301;
    ctx->b = 0xefcdab89;
    ctx->c = 0x98badcfe;
    ctx->d = 0x10325476;

    ctx->lo = 0;
    ctx->hi = 0;
}

void MD5_Update(MD5_CTX* ctx, const void* data, unsigned long size) {
    MD5_u32plus saved_lo;
    unsigned long used, available;

    saved_lo = ctx->lo;
    if ((ctx->lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
        ctx->hi++;
    ctx->hi += size >> 29;

    used = saved_lo & 0x3f;

    if (used) {
        available = 64 - used;

        if (size < available) {
            memcpy(&ctx->buffer[used], data, size);
            return;
        }

        memcpy(&ctx->buffer[used], data, available);
        data = (const unsigned char*)data + available;
        size -= available;
        body(ctx, ctx->buffer, 64);
    }

    if (size >= 64) {
        data = body(ctx, data, size & ~(unsigned long)0x3f);
        size &= 0x3f;
    }

    memcpy(ctx->buffer, data, size);
}

#define OUT(dst, src)                                                          \
    (dst)[0] = (unsigned char)(src);                                           \
    (dst)[1] = (unsigned char)((src) >> 8);                                    \
    (dst)[2] = (unsigned char)((src) >> 16);                                   \
    (dst)[3] = (unsigned char)((src) >> 24);

void MD5_Final(unsigned char* result, MD5_CTX* ctx) {
    unsigned long used, available;

    used = ctx->lo & 0x3f;

    ctx->buffer[used++] = 0x80;

    available = 64 - used;

    if (available < 8) {
        memset(&ctx->buffer[used], 0, available);
        body(ctx, ctx->buffer, 64);
        used = 0;
        available = 64;
    }

    memset(&ctx->buffer[used], 0, available - 8);

    ctx->lo <<= 3;
    OUT(&ctx->buffer[56], ctx->lo)
    OUT(&ctx->buffer[60], ctx->hi)

    body(ctx, ctx->buffer, 64);

    OUT(&result[0], ctx->a)
    OUT(&result[4], ctx->b)
    OUT(&result[8], ctx->c)
    OUT(&result[12], ctx->d)

    memset(ctx, 0, sizeof(*ctx));
}

#endif // HAVE_OPENSSL
